rust/src/module_expressions.rs

```diff
@@ -0,0 +1,196 @@
+use crate::errors::{GrimpError, GrimpResult};
+use const_format::formatcp;
+use itertools::Itertools;
+use lazy_static::lazy_static;
+use regex::Regex;
+use std::fmt::Display;
+use std::str::FromStr;
+
+lazy_static! {
+    static ref MODULE_EXPRESSION_PATTERN: Regex =
+        Regex::new(r"^(\w+|\*{1,2})(\.(\w+|\*{1,2}))*$").unwrap();
+}
+
+/// A module expression is used to refer to sets of modules.
+///
+/// - `*` stands in for a module name, without including subpackages.
+/// - `**` includes subpackages too.
+#[derive(Debug, Clone)]
+pub struct ModuleExpression {
+    expression: String,
+    pattern: Regex,
+}
+
+impl Display for ModuleExpression {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.expression)
+    }
+}
+
+impl AsRef<str> for ModuleExpression {
+    fn as_ref(&self) -> &str {
+        &self.expression
+    }
+}
+
+impl FromStr for ModuleExpression {
+    type Err = GrimpError;
+
+    fn from_str(expression: &str) -> Result<Self, Self::Err> {
+        if !MODULE_EXPRESSION_PATTERN.is_match(expression) {
+            return Err(GrimpError::InvalidModuleExpression(expression.to_owned()));
+        }
+
+        for (part, next_part) in expression.split(".").tuple_windows() {
+            match (part, next_part) {
+                ("*", "**") | ("**", "*") | ("**", "**") => {
+                    return Err(GrimpError::InvalidModuleExpression(expression.to_owned()))
+                }
+                _ => {}
+            }
+        }
+
+        Ok(Self {
+            expression: expression.to_owned(),
+            pattern: Self::create_pattern(expression)?,
+        })
+    }
+}
+
+const MODULE_NAME_REGEX: &str = r"[^\.]+";
+const MULTIPLE_MODULE_NAMES_REGEX: &str =
```

Comment:
🐼  Maybe `MODULE_NAME_PATTERN` is slightly more correct, since it's not technically a `Regex`.

Reply:
Good point, I'll change.